wasi/src/main/java/com/dylibso/chicory/wasi/WasiPreview1.java

```diff
@@ -1339,9 +1338,122 @@ public int pathUnlinkFile(int dirFd, @Buffer String rawPath) {
+                    if (flagSet(flags, WasiSubClockFlags.SUBSCRIPTION_CLOCK_ABSTIME)) {
+                        timeout -= clockTime(clockId);
+                    }
+                    clockEvents.add(Map.entry(timeout, userData));
+                    break;
+                case WasiEventType.FD_READ:
+                case WasiEventType.FD_WRITE:
+                    int fd = memory.readInt(inPtr);
+                    if (fd < 0) {
+                        return wasiResult(WasiErrno.EBADF);
+                    }
+                    var descriptor = descriptors.get(fd);
+                    if (descriptor instanceof InStream && eventType == WasiEventType.FD_READ) {
+                        readEvents.add(Map.entry((InStream) descriptor, userData));
+                    } else if (descriptor instanceof OutStream
+                            && eventType == WasiEventType.FD_WRITE) {
+                        // assume output streams are always writable
+                        writeEvent(memory, outPtr, userData, eventType, WasiErrno.ESUCCESS);
+                        outPtr += 32;
+                        nevents++;
+                    } else {
+                        WasiErrno errno;
+                        if (descriptor == null) {
+                            errno = WasiErrno.EBADF;
+                        } else if (descriptor instanceof OpenFile) {
+                            // per specification: this event always triggers for regular files
+                            errno = WasiErrno.ESUCCESS;
+                        } else {
+                            errno = WasiErrno.ENOTSUP;
+                        }
+                        writeEvent(memory, outPtr, userData, eventType, errno);
+                        outPtr += 32;
+                        nevents++;
+                    }
+                    break;
+                default:
+                    return wasiResult(WasiErrno.EINVAL);
+            }
+            inPtr += 32;
+        }
+
+        long minTimeout = clockEvents.stream().mapToLong(Entry::getKey).min().orElse(0);
+        long start = System.nanoTime();
+        while (true) {
+            for (var entry : readEvents) {
+                InStream stream = entry.getKey();
+                long userData = entry.getValue();
+                try {
+                    int available = stream.available();
+                    if (available <= 0) {
+                        continue;
+                    }
+                    writeEvent(memory, outPtr, userData, WasiEventType.FD_READ, WasiErrno.ESUCCESS);
+                    memory.writeLong(outPtr + 16, available);
+                    outPtr += 32;
+                    nevents++;
+                } catch (IOException e) {
+                    writeEvent(memory, outPtr, userData, WasiEventType.FD_READ, WasiErrno.EIO);
+                    outPtr += 32;
+                    nevents++;
+                }
+            }
+            if (nevents > 0) {
```

Comment:
What's the reference implementation you started from?

I cannot find this condition in the [wazero implementation](https://github.com/tetratelabs/wazero/blob/610c202ec48f3a7c729f2bf11707330127ab3689/imports/wasi_snapshot_preview1/poll.go#L161).

Reply:
I wrote this myself based on the specification. The wazero implementation seems incorrect as it writes out all clock events, when per the specification it should only be the clocks that reached the timeout:

> The time value of clock `subscription_clock::id` has reached timestamp `subscription_clock::timeout`.

This condition is actually incorrect, as it will prevent writing out clocks if there were any triggered descriptors. Instead, I'll change loop to skip sleeping if it already wrote events.

Reply:
> I wrote this myself based on the specification.

Can you add a comment with the relevant (perma-) link?

Reply:
This is just the specification for WASI. We could add links for all the functions as a separate commit if you think that is valuable.